### PR TITLE
Torch filters

### PR DESCRIPTION
**Pull request by jorshi**
Starting to make some differentiable filters. Moving average and FIR LPF are included in this PR. In the future I think we are going to want to compute convolutions in the frequency domain as opposed to doing it using torch conv1d as that will likely be more efficient, this is what Engel et al. did in the DDSP work. But I will leave that for a future change since this is working for now. Not including fuzzy tests since the numpy convolution operation and torch.nn.functional.conv1d operations aren't quite the same.

### CODE REVIEW

File: src/ddspdrum/torchmodule.py [hunk 1]
```diff
@@ -485,3 +485,146 @@ def _forward(self, control_in: T, audio_in: T) -> T:
...
+        audio (T)  :   audio samples to filter
+        """
+
+        impulse = self.windowed_sinc(self.p("cutoff"), self.p("length"))
+        impulse = torch.reshape(impulse, (1, 1, impulse.size()[0]))
+        audio_resized = torch.reshape(audio_in, (1, 1, audio_in.size()[0]))
+        y = nn.functional.conv1d(audio_resized, impulse, padding=int(self.p("length") / 2))
+        return y[0][0]
+
+    def windowed_sinc(self, cutoff: T, filter_length: T) -> T:
```
Comment (jorshi) [inline src/ddspdrum/torchmodule.py:545]: I don't love the code in this method, but it was challenging to make this differentiable in a way that felt really clean. The big obstacle that I encountered in both of the filters was making the filter length differentiable.

File: tests/test_torchmodule.py [hunk 2]
```diff
@@ -107,6 +107,7 @@ def test_TorchSineVCO(self):
         )
 
     def test_TorchFmVCO(self):
+        # Is this correct?
         numpymod = numpymodule.FmVCO()
         torchmod = torchmodule.TorchSineVCO()
```
Comment (jorshi) [inline tests/test_torchmodule.py:112]: Not really part of this PR, but wanted to flag this. should this be comparing FmVCO and TorchFmVCO -- why is this test working?

File: tests/test_torchmodule.py [hunk 3]
```diff
@@ -117,6 +118,18 @@ def test_TorchFmVCO(self):
             }
         )
 
+    # Fuzzy test doesn't work for filtering
```
Comment (jorshi) [inline tests/test_torchmodule.py:121]: Going to pull this out.

File: src/ddspdrum/torchmodule.py [hunk 4]
```diff
@@ -485,3 +485,146 @@ def _forward(self, control_in: T, audio_in: T) -> T:
 
         audio_in = fix_length(audio_in, len(control_in))
         return control_in * audio_in
+
+
+class TorchFIR(TorchSynthModule):
```
Comment (jorshi) [inline src/ddspdrum/torchmodule.py:490]: Considering renaming this -- FIRLowPass or something similar that is more specific

File: tests/test_torchmodule.py (hunk 2 above)
Comment (turian) [reply tests/test_torchmodule.py:112]: hmmm @maxsolomonhenry ? thoughts?

File: src/ddspdrum/torchmodule.py [hunk 5]
```diff
@@ -485,3 +485,154 @@ def _forward(self, control_in: T, audio_in: T) -> T:
...
+        TODO: Cutoff frequency modulation, if there is an efficient way to do it
+
+        Parameters
+        ----------
+
+        audio (T)  :   audio samples to filter
+        """
+
+        impulse = self.windowed_sinc(self.p("cutoff"), self.p("length"))
+        impulse = torch.reshape(impulse, (1, 1, impulse.size()[0]))
```
Comment (turian) [inline src/ddspdrum/torchmodule.py:540]: reshape looses gradients I think? You probably want view?

Are you able to backprop through this?

File: src/ddspdrum/torchmodule.py (hunk 1 above)
Comment (turian) [reply src/ddspdrum/torchmodule.py:549]: Cool, might also be worth comparing to https://github.com/mravanelli/SincNet/blob/788fc3cd42e5c3172c5a8525b3a872eb7205c8f9/dnn_models.py and https://github.com/adefossez/julius

File: tests/test_torchmodule.py (hunk 2 above)
Comment (maxsolomonhenry) [reply tests/test_torchmodule.py:112]: Hah. Whoops. Would you mind tweaking it? Odd how it didn't seem to trigger any failure...

File: src/ddspdrum/torchmodule.py (hunk 1 above)
Comment (jorshi) [reply src/ddspdrum/torchmodule.py]: Nice, the julius code was helpful for cleaning up the windowed sinc implementation a bit. That FFT based convolution implementation is sweet, would be cool to explore that in the future.

File: tests/test_torchmodule.py (hunk 2 above)
Comment (jorshi) [reply tests/test_torchmodule.py:112]: Consider it tweaked. That is odd that it didn't trigger any failure?

File: src/ddspdrum/torchmodule.py (hunk 5 above)
Comment (jorshi) [reply src/ddspdrum/torchmodule.py:540]: Yes able to backprop through this -- is view better though?

Comment (jorshi) [reply src/ddspdrum/torchmodule.py]: So I guess reshape will change the view if possible, otherwise it may copy the data -- so I think view is better here, made this change.

File: examples/examples.py [hunk 6]
```diff
@@ -554,3 +554,112 @@ def stft_plot(signal, sample_rate=SAMPLE_RATE):
...
+# Create some noise to filter
+duration = 2
+noise = torch.tensor(np.random.rand(int(44100 * duration)) * 2 - 1).float()
+stft_plot(noise.detach().numpy())
+# -
+
+# **Moving Average Filter**
+#
+# A moving average filter is a simple finite impulse response (FIR) filter that calculates that value of a sample by taking the average of M input samples at a time. The filter_length defines how many samples M to include in the average.
+
```
Comment (maxsolomonhenry) [inline examples/examples.py:573]: As I mentioned in the fb chat, the output here looks like and sounds like comb filter and not an MA filter. Not sure about the guts of it; and maybe I'm off... but would expect an LP type behaviour.

I might be totally off base. Lmk.
